Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -326,6 +344,55 @@ theorem bitwise_swap {f : Bool → Bool → Bool} :
     simp [bitwise_of_ne_zero, ih]
 #align nat.bitwise_swap Nat.bitwise_swap
 
+theorem testBit_two_pow_mul_add {n b w i} (h: i < w) :
+    Nat.testBit (2 ^ w * b + n) i = Nat.testBit n i := by
+  simp only [testBit, shiftRight_eq_div_pow, bodd_eq_bodd_iff]
+  rw [Nat.add_div_of_dvd_right (by simp [Dvd.dvd.mul_right, pow_dvd_pow, le_of_lt h]), add_mod]
+  have h1: (2 ^ w / 2 ^ i) % 2 = 0 := by
+    rw [← Nat.dvd_iff_mod_eq_zero]
+    use 2 ^ (w - (i + 1))
+    rw [Nat.pow_div (by linarith) (by decide), mul_comm, ← pow_succ 2 _, succ_eq_add_one]
+    simp [← Nat.sub_add_comm, succ_le_of_lt h]
+  simp [mul_comm, Nat.mul_div_assoc b (pow_dvd_pow 2 (le_of_lt h)), mul_mod, h1]
+
+theorem testBit_two_pow_mul_toNat_add {n w b} (h: n < 2 ^ w) :
+    testBit (2 ^ w * b.toNat + n) w = b := by
+  simp only [testBit, shiftRight_eq_div_pow]
+  rw [Nat.add_div_of_dvd_right (Dvd.intro _ rfl), Nat.div_eq_of_lt h, add_zero]
+  cases' b <;> simp
+
+/-- Generic method to create a natural number by appending bits tail-recursively.
+It takes a boolean function `f` on each bit and `z` the starting point and the number of bits `i`.
+It is almost always specialized with `z = 0` and `i = w`; the length of the binary representation.
+Note that `ofBits f z i = 2 ^ i * z + ofBits f 0 i` which we prove next.
+This is an alternative to using `List`. It will be used for bitadd, bitneg, bitmul etc.-/
+def ofBits (f : Nat → Bool) (z : Nat) : Nat → Nat
```

Comment:
Because `z` is necessarily needed in the definition. At each step, it gets populated by the carry-out bit and then gets appended to the binary representation.

Reply:
How about
```lean
def ofBits (f : Nat → Bool) : Nat → Nat :=
  go 0
  where
    go (z : Nat) : Nat → Nat
    | 0 => z
    | i + 1 => go (z.bit (f i)) i
```
This conveys properly that `z` is just an internal accumulator, and that the public API should only ever have `z = 0`

Reply:
I moved the proves to this version of ofBits.

Reply:
Thanks!